pkg/spire/verify.go

```diff
@@ -0,0 +1,234 @@
+}
+
+func getSVID(resultMap map[string]v1beta1.PipelineResourceResult) (*x509.Certificate, error) {
+	svid, ok := resultMap[KeySVID]
+	if !ok {
+		return nil, errors.New("no SVID found")
+	}
+	svidValue, err := getResultValue(svid)
+	if err != nil {
+		return nil, err
+	}
+	block, _ := pem.Decode([]byte(svidValue))
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("invalid SVID: %w", err)
+	}
+	return cert, nil
+}
+
+func getTrustBundle(ctx context.Context, client *workloadapi.Client) (*x509.CertPool, error) {
+	x509set, err := client.FetchX509Bundles(ctx)
+	if err != nil {
+		return nil, err
+	}
+	x509Bundle := x509set.Bundles()
+	if err != nil {
+		return nil, err
+	}
+	trustPool := x509.NewCertPool()
+	for _, c := range x509Bundle[0].X509Authorities() {
+		trustPool.AddCert(c)
+	}
+	return trustPool, nil
+}
+
+func getFullPath(tr *v1beta1.TaskRun) string {
+	// URI:spiffe://example.org/ns/default/taskrun/cache-image-pipelinerun-r4r22-fetch-from-git
+	return fmt.Sprintf("/ns/%s/taskrun/%s", tr.Namespace, tr.Name)
+}
+
+func verifyCertURI(cert *x509.Certificate, tr *v1beta1.TaskRun, trustDomain string) error {
+	path := getFullPath(tr)
+	if len(cert.URIs) > 0 {
+		if cert.URIs[0].Host != trustDomain {
+			return fmt.Errorf("cert uri: %s does not match trust domain: %s", cert.URIs[0].Host, trustDomain)
+		}
+		if cert.URIs[0].Path != path {
+			return fmt.Errorf("cert uri: %s does not match taskrun: %s", cert.URIs[0].Path, path)
+		}
+	} else {
+		return fmt.Errorf("cert uri missing for taskrun: %s", tr.Name)
+	}
+
+	return nil
+}
+
+func verifyCertificateTrust(cert *x509.Certificate, rootCertPool *x509.CertPool) error {
+	verifyOptions := x509.VerifyOptions{
+		Roots: rootCertPool,
+	}
+	chains, err := cert.Verify(verifyOptions)
+	if len(chains) == 0 || err != nil {
+		return errors.New("cert cannot be verified by provided roots")
+	}
+	return nil
+}
+
+func verifyManifest(results map[string]v1beta1.PipelineResourceResult) error {
+	manifest, ok := results[KeyResultManifest]
+	if !ok {
+		return errors.New("no manifest found in results")
+	}
+	manifestValue, err := getResultValue(manifest)
+	if err != nil {
+		return err
+	}
+	s := strings.Split(manifestValue, ",")
+	for _, key := range s {
+		_, found := results[key]
+		if key != "" && !found {
+			return fmt.Errorf("no result found for %s but is part of the manifest %s", key, manifestValue)
+		}
+	}
+	return nil
+}
+
+func verifyResult(pub interface{}, key string, results map[string]v1beta1.PipelineResourceResult) error {
+	signature, ok := results[key+KeySignatureSuffix]
+	if !ok {
+		return fmt.Errorf("no signature found for %s", key)
+	}
+	sigValue, err := getResultValue(signature)
+	if err != nil {
+		return err
+	}
+	resultValue, err := getResultValue(results[key])
+	if err != nil {
+		return err
+	}
+	return verifySignature(pub, sigValue, resultValue)
+}
+
+func verifySignature(pub interface{}, signature string, value string) error {
```

Comment:
This doesn't make a functional difference, but should we use this for better type hinting from the func args? (also for verifyResult)

```suggestion
func verifySignature(pub crypto.PublicKey, signature string, value string) error {
```

Reply:
Agreed. Changed.